File: QSNode/qsettlement/chain/consensus/pq_signer.go
package consensus

import (
	"crypto/sha3"
	"errors"
	"fmt"
	"log"

	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/ethereum/go-ethereum/rlp"

	// Import Dilithium from QSmart (path adjusted based on actual module layout)
	// "github.com/mchawda/qsmart/pqcrypto/dilithium"
)

// PQSigner handles post-quantum cryptographic signing operations
type PQSigner struct {
	algorithm   string // "dilithium2", "dilithium3", etc.
	privateKey  []byte // PQ private key
	publicKey   []byte // PQ public key
	address     common.Address // Derived address from PQ public key
}

// NewPQSigner creates a new PQ signer instance
func NewPQSigner(algo string, privKey []byte, pubKey []byte) (*PQSigner, error) {
	if algo != "dilithium2" && algo != "dilithium3" {
		return nil, fmt.Errorf("unsupported PQ algorithm: %s", algo)
	}

	if len(privKey) == 0 || len(pubKey) == 0 {
		return nil, errors.New("private key and public key cannot be empty")
	}

	// Derive address from PQ public key using Keccak256
	// address = last 20 bytes of keccak256(pubKey)
	hash := sha3.NewLegacyKeccak256()
	hash.Write(pubKey)
	hashBytes := hash.Sum(nil)
	address := common.BytesToAddress(hashBytes[12:32]) // Take last 20 bytes

	return &PQSigner{
		algorithm:  algo,
		privateKey: privKey,
		publicKey:  pubKey,
		address:    address,
	}, nil
}

// GetAddress returns the derived address
func (ps *PQSigner) GetAddress() common.Address {
	return ps.address
}

// GetPublicKey returns the public key
func (ps *PQSigner) GetPublicKey() []byte {
	return ps.publicKey
}

// HashHeaderSansSig creates a hash of the block header without PQ signature fields
// This is the data that will be signed with Dilithium2
func HashHeaderSansSig(header interface{}) ([]byte, error) {
	// Create a copy of header without PQ fields
	// In actual implementation, this would work with the proto-generated struct
	
	// For now, we demonstrate the concept:
	// Encode header without pq_sig, pq_agg_sig fields
	headerBytes, err := rlp.EncodeToBytes(header)
	if err != nil {
		return nil, fmt.Errorf("failed to encode header: %w", err)
	}

	// Compute Keccak256 hash
	hash := sha3.NewLegacyKeccak256()
	hash.Write(headerBytes)
	return hash.Sum(nil), nil
}

// ProposerSignHeader signs a block header with Dilithium2
// This is called by the proposer to create the pq_sig field
func (ps *PQSigner) ProposerSignHeader(headerHash []byte) ([]byte, error) {
	if ps.algorithm != "dilithium2" && ps.algorithm != "dilithium3" {
		return nil, fmt.Errorf("unsupported algorithm for signing: %s", ps.algorithm)
	}

	// In production: dilithium.Sign(ps.privateKey, headerHash)
	// For now, we log the operation (actual signing will be done via QSmart library)
	log.Printf("[PQ] Signing header with %s, hash: %x\n", ps.algorithm, headerHash[:8])

	// TODO: Replace with actual Dilithium2 signing from QSmart
	// sig := dilithium.Sign(ps.privateKey, headerHash)

	// Placeholder: return mock signature (in practice, use QSmart library)
	// Dilithium2 signatures are typically ~2.7-3.0 KB
	mockSig := make([]byte, 2800) // Mock ~2.8KB signature
	copy(mockSig, []byte("DILITHIUM2_SIG_PLACEHOLDER_"))
	return mockSig, nil
}

// VerifyPQSignature verifies a PQ signature using Dilithium2
// This is called by validators to verify the proposer's signature
func (ps *PQSigner) VerifyPQSignature(pubKey []byte, headerHash []byte, signature []byte) bool {
	if len(pubKey) == 0 || len(signature) == 0 {
		log.Println("[PQ] Verification failed: empty public key or signature")
		return false
	}

	// In production: dilithium.Verify(pubKey, headerHash, signature)
	log.Printf("[PQ] Verifying %s signature, pubkey len: %d, sig len: %d\n", 
		ps.algorithm, len(pubKey), len(signature))

	// TODO: Replace with actual Dilithium2 verification from QSmart
	// return dilithium.Verify(pubKey, headerHash, signature)

	// Placeholder: accept mock signatures for now
	return len(signature) > 0 && len(pubKey) > 0
}

// VoteDigest creates a hash of a consensus vote for signing
// This is used for both proposal votes and commit votes
func VoteDigest(round uint64, height uint64, blockID string) []byte {
	// Encode vote components
	voteData := struct {
		Round   uint64
		Height  uint64
		BlockID string
	}{round, height, blockID}

	data, err := rlp.EncodeToBytes(voteData)
	if err != nil {
		log.Printf("Failed to encode vote: %v\n", err)
		return nil
	}

	// Hash the vote data
	hash := sha3.NewLegacyKeccak256()
	hash.Write(data)
	return hash.Sum(nil)
}

// SignVote signs a consensus vote with Dilithium2
// Used in BFT consensus to sign votes/commits
func (ps *PQSigner) SignVote(round uint64, height uint64, blockID string) ([]byte, error) {
	voteHash := VoteDigest(round, height, blockID)
	if voteHash == nil {
		return nil, errors.New("failed to create vote digest")
	}

	// TODO: Replace with actual Dilithium2 signing
	log.Printf("[PQ] Signing vote for round %d, height %d\n", round, height)

	// Placeholder signature
	mockSig := make([]byte, 2800)
	copy(mockSig, []byte("DILITHIUM2_VOTE_"))
	return mockSig, nil
}

// VerifyVote verifies a consensus vote signature
func VerifyVote(pubKey []byte, round uint64, height uint64, blockID string, signature []byte) bool {
	voteHash := VoteDigest(round, height, blockID)
	if voteHash == nil {
		return false
	}

	// TODO: Replace with actual Dilithium2 verification
	log.Printf("[PQ] Verifying vote signature, sig len: %d\n", len(signature))
	return len(signature) > 0 && len(pubKey) > 0
}

// ValidatorRegistry tracks validator PQ keys and enables key rotation
type ValidatorRegistry struct {
	validators map[string]*ValidatorInfo
}

// ValidatorInfo stores information about a validator
type ValidatorInfo struct {
	Address     common.Address
	PQPublicKey []byte
	Algorithm   string
	Power       uint64
	Active      bool
	EpochAdded  uint64
	EpochExpired uint64
}

// NewValidatorRegistry creates a new validator registry
func NewValidatorRegistry() *ValidatorRegistry {
	return &ValidatorRegistry{
		validators: make(map[string]*ValidatorInfo),
	}
}

// AddValidator adds a new validator to the registry
func (vr *ValidatorRegistry) AddValidator(addr common.Address, pubKey []byte, algo string, power uint64, epoch uint64) error {
	if len(pubKey) == 0 {
		return errors.New("public key cannot be empty")
	}

	key := addr.Hex()
	vr.validators[key] = &ValidatorInfo{
		Address:     addr,
		PQPublicKey: pubKey,
		Algorithm:   algo,
		Power:       power,
		Active:      true,
		EpochAdded:  epoch,
		EpochExpired: 0,
	}

	log.Printf("[ValidatorRegistry] Added validator %s with %s key (power: %d)\n", addr.Hex(), algo, power)
	return nil
}

// RotateValidatorKey rotates a validator's PQ key to a new one (epoch-based)
func (vr *ValidatorRegistry) RotateValidatorKey(addr common.Address, newPubKey []byte, newAlgo string, epoch uint64) error {
	key := addr.Hex()
	validator, exists := vr.validators[key]
	if !exists {
		return fmt.Errorf("validator %s not found", addr.Hex())
	}

	// Mark old key as expired
	validator.EpochExpired = epoch

	// Add new validator entry with rotated key
	vr.validators[key+"_new"] = &ValidatorInfo{
		Address:     addr,
		PQPublicKey: newPubKey,
		Algorithm:   newAlgo,
		Power:       validator.Power,
		Active:      true,
		EpochAdded:  epoch,
		EpochExpired: 0,
	}

	log.Printf("[ValidatorRegistry] Rotated key for validator %s at epoch %d\n", addr.Hex(), epoch)
	return nil
}

// GetActiveValidator retrieves active validator info
func (vr *ValidatorRegistry) GetActiveValidator(addr common.Address) (*ValidatorInfo, error) {
	key := addr.Hex()
	if validator, exists := vr.validators[key]; exists && validator.Active && validator.EpochExpired == 0 {
		return validator, nil
	}

	// Try to find most recent validator key
	if validator, exists := vr.validators[key+"_new"]; exists && validator.Active {
		return validator, nil
	}

	return nil, fmt.Errorf("no active validator found for %s", addr.Hex())
}

// GetAllActiveValidators returns all currently active validators
func (vr *ValidatorRegistry) GetAllActiveValidators() []*ValidatorInfo {
	var active []*ValidatorInfo
	for _, v := range vr.validators {
		if v.Active && v.EpochExpired == 0 {
			active = append(active, v)
		}
	}
	return active
}

File: QSNode/sdk/typescript/pq-signer.ts
/**
 * @qsmart/pq-signer
 * 
 * TypeScript SDK for signing Quantum-EVM (EIP-2718 type 0x79) transactions
 * with post-quantum cryptography (Dilithium2).
 * 
 * Usage:
 *   const signer = new PQSigner(dilithium2PrivateKey);
 *   const tx = { chainId: 9357, nonce: 0, ... };
 *   const signedTx = await signer.signTransaction(tx);
 */

import { keccak256 as ethersKeccak256, RLP } from "ethers";

// Types for PQ transaction
export interface PQTransactionRequest {
  chainId: number;
  nonce: number;
  maxPriorityFeePerGas: bigint;
  maxFeePerGas: bigint;
  gas: number;
  to?: string; // Recipient address
  value: bigint;
  data?: string; // Hex-encoded data
  accessList?: AccessTuple[];
}

export interface AccessTuple {
  address: string;
  storageKeys: string[];
}

export interface PQSignedTransaction {
  type: 0x79;
  chainId: number;
  nonce: number;
  maxPriorityFeePerGas: bigint;
  maxFeePerGas: bigint;
  gas: number;
  to?: string;
  value: bigint;
  data?: string;
  accessList?: AccessTuple[];
  pqSigAlgo: number; // 0x01 = Dilithium2
  pqPublicKey: Uint8Array;
  pqSignature: Uint8Array;
  from: string; // Derived address
  hash: string; // Tx hash
}

/**
 * PQSigner handles post-quantum transaction signing with Dilithium2
 */
export class PQSigner {
  private algorithm: string;
  private privateKey: Uint8Array;
  private publicKey: Uint8Array;
  private address: string;

  constructor(
    algorithm: string = "dilithium2",
    privateKey: Uint8Array,
    publicKey: Uint8Array
  ) {
    if (algorithm !== "dilithium2" && algorithm !== "dilithium3") {
      throw new Error(`Unsupported PQ algorithm: ${algorithm}`);
    }

    this.algorithm = algorithm;
    this.privateKey = privateKey;
    this.publicKey = publicKey;
    this.address = this.deriveAddress();
  }

  /**
   * Derives the Ethereum-style address from the PQ public key
   * address = last 20 bytes of keccak256(pubkey)
   */
  private deriveAddress(): string {
    const hash = ethersKeccak256(this.publicKey);
    // Take last 20 bytes (40 hex chars) as address
    return "0x" + hash.slice(-40).toLowerCase();
  }

  /**
   * Gets the signer's address
   */
  getAddress(): string {
    return this.address;
  }

  /**
   * Gets the PQ public key
   */
  getPublicKey(): Uint8Array {
    return this.publicKey;
  }

  /**
   * Computes the signing hash for a transaction
   * This is the data that will be signed with Dilithium2
   */
  private computeSigningHash(tx: PQTransactionRequest): Uint8Array {
    // Prepare data to be signed: exclude signature fields
    const sighashParts = [
      tx.chainId,
      tx.nonce,
      tx.maxPriorityFeePerGas.toString(16),
      tx.maxFeePerGas.toString(16),
      tx.gas,
      tx.to || "0x",
      tx.value.toString(16),
      tx.data || "0x",
      tx.accessList || [],
      0x01, // pqSigAlgo (Dilithium2)
      Array.from(this.publicKey), // pqPublicKey
    ];

    // Encode to RLP
    const encoded = RLP.encode(sighashParts);

    // Hash with Keccak256
    const hash = ethersKeccak256(encoded);

    // Convert hex string to Uint8Array
    const hashStr = hash.slice(2); // Remove '0x'
    const bytes = new Uint8Array(hashStr.length / 2);
    for (let i = 0; i < hashStr.length; i += 2) {
      bytes[i / 2] = parseInt(hashStr.substr(i, 2), 16);
    }

    return bytes;
  }

  /**
   * Signs a transaction with Dilithium2
   * In production, this uses the actual Dilithium2 library from @qsmart/pq-crypto
   */
  async signTransaction(tx: PQTransactionRequest): Promise<PQSignedTransaction> {
    const signingHash = this.computeSigningHash(tx);

    // TODO: Replace with actual Dilithium2 signing from @qsmart/pq-crypto
    // const signature = await dilithium.sign(this.privateKey, signingHash);

    // Placeholder: create a mock signature for demonstration
    // In production: use actual Dilithium2 signing library
    const signature = this.createMockSignature(signingHash);

    // Compute transaction hash
    const txHash = this.computeTransactionHash(tx, signature);

    return {
      type: 0x79,
      chainId: tx.chainId,
      nonce: tx.nonce,
      maxPriorityFeePerGas: tx.maxPriorityFeePerGas,
      maxFeePerGas: tx.maxFeePerGas,
      gas: tx.gas,
      to: tx.to,
      value: tx.value,
      data: tx.data,
      accessList: tx.accessList,
      pqSigAlgo: 0x01, // Dilithium2
      pqPublicKey: this.publicKey,
      pqSignature: signature,
      from: this.address,
      hash: txHash,
    };
  }

  /**
   * Creates a mock Dilithium2 signature for demonstration
   * In production, this is replaced by actual cryptographic signing
   */
  private createMockSignature(data: Uint8Array): Uint8Array {
    // Dilithium2 signatures are typically ~2.7-3.0 KB
    const mockSize = 2800;
    const signature = new Uint8Array(mockSize);

    // Fill with deterministic but non-trivial data
    for (let i = 0; i < mockSize; i++) {
      signature[i] = (data[i % data.length] + i) % 256;
    }

    return signature;
  }

  /**
   * Computes the transaction hash
   */
  private computeTransactionHash(
    tx: PQTransactionRequest,
    signature: Uint8Array
  ): string {
    // Encode the full signed transaction
    const txData = [
      tx.chainId,
      tx.nonce,
      tx.maxPriorityFeePerGas.toString(16),
      tx.maxFeePerGas.toString(16),
      tx.gas,
      tx.to || "0x",
      tx.value.toString(16),
      tx.data || "0x",
      tx.accessList || [],
      0x01, // pqSigAlgo
      Array.from(this.publicKey),
      Array.from(signature),
    ];

    // Encode to RLP with type prefix
    const encoded = RLP.encode(txData);
    const txBytes = new Uint8Array(1 + encoded.length);
    txBytes[0] = 0x79; // Type 0x79
    txBytes.set(new Uint8Array(encoded), 1);

    // Hash the result
    const hash = ethersKeccak256(txBytes);
    return hash;
  }
}

/**
 * Generates a new Dilithium2 keypair
 * In production, this uses the actual Dilithium2 library
 */
export async function generateKeyPair(): Promise<{
  publicKey: Uint8Array;
  privateKey: Uint8Array;
}> {
  // TODO: Replace with actual Dilithium2 key generation from @qsmart/pq-crypto
  // const { pk, sk } = await dilithium.generateKeyPair();

  // Placeholder: generate mock keys
  const pubKey = new Uint8Array(1184); // Dilithium2 pubkey ~1184 bytes
  const privKey = new Uint8Array(2400); // Dilithium2 privkey ~2400 bytes

  // Fill with random-like data
  crypto.getRandomValues(pubKey);
  crypto.getRandomValues(privKey);

  return {
    publicKey: pubKey,
    privateKey: privKey,
  };
}

/**
 * Verifies a PQ transaction signature (for validation)
 */
export function verifyPQTransaction(tx: PQSignedTransaction): boolean {
  if (!tx.pqPublicKey || tx.pqPublicKey.length === 0) {
    console.error("Public key is empty");
    return false;
  }

  if (!tx.pqSignature || tx.pqSignature.length === 0) {
    console.error("Signature is empty");
    return false;
  }

  if (tx.pqSigAlgo !== 0x01) {
    console.error(`Unsupported signature algorithm: ${tx.pqSigAlgo}`);
    return false;
  }

  // TODO: Actual verification with Dilithium2
  // return await dilithium.verify(tx.pqPublicKey, signingHash, tx.pqSignature);

  console.log("[PQSigner] Verifying PQ transaction");
  console.log(`  Public key size: ${tx.pqPublicKey.length} bytes`);
  console.log(`  Signature size: ${tx.pqSignature.length} bytes`);
  console.log(`  Algorithm: Dilithium2 (${tx.pqSigAlgo})`);

  // Placeholder: accept for now
  return tx.pqSignature.length > 100;
}

/**
 * Encodes a signed PQ transaction to 0x-prefixed hex string
 */
export function encodeSignedTransaction(tx: PQSignedTransaction): string {
  const txData = [
    tx.chainId,
    tx.nonce,
    tx.maxPriorityFeePerGas.toString(16),
    tx.maxFeePerGas.toString(16),
    tx.gas,
    tx.to || "0x",
    tx.value.toString(16),
    tx.data || "0x",
    tx.accessList || [],
    tx.pqSigAlgo,
    Array.from(tx.pqPublicKey),
    Array.from(tx.pqSignature),
  ];

  const encoded = RLP.encode(txData);
  const bytes = new Uint8Array(1 + encoded.length);
  bytes[0] = 0x79; // Type
  bytes.set(new Uint8Array(encoded), 1);

  // Convert to hex string
  let hexStr = "0x";
  for (const byte of bytes) {
    hexStr += byte.toString(16).padStart(2, "0");
  }

  return hexStr;
}

export default PQSigner;
